File: C/2024/day02.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

bool safe(int *row, size_t len)
{
    bool safe = true;
    bool inc  = true;

    for (size_t i = 0; i < len - 1; ++i)
    {
        // Set increment on first pass
        if (i == 0)
            inc = (row[i + 1] > row[i]);

        // Check for increment swap
        if ((inc && row[i + 1] < row[i]) || (!inc && row[i + 1] > row[i]))
        {
            safe = false;
            break;
        }

        // Check bounds
        if (abs(row[i + 1] - row[i]) < 1 || abs(row[i + 1] - row[i]) > 3)
        {
            safe = false;
            break;
        }
    }

    return safe;
}

int *tokenize(char *line, size_t *len)
{
    size_t i   = 0;
    int   *arr = calloc(++i, sizeof(int));

    char *ptr;
    char *token = strtok_r(line, " ", &ptr);
    while (token)
    {
        // Get number
        long num   = strtol(token, NULL, 10);
        arr[i - 1] = (int)num;

        // Resize and get next
        arr   = realloc(arr, (++i) * sizeof(int));
        token = strtok_r(NULL, " ", &ptr);
    }

    *len = i - 1;
    return arr;
}

int *pop(int *row, size_t len, size_t idx)
{
    // Make a copy
    int *copy = calloc(len - 1, sizeof(int));

    // Copy
    memcpy(copy, row, idx * sizeof(int));
    memcpy(&copy[idx], &row[idx + 1], (len - idx - 1) * sizeof(int));

    return copy;
}

int main(int argc, char **argv)
{
    // Args
    if (argc != 2)
    {
        fprintf(stderr, "Usage: %s <filename>\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    // Get filename
    FILE *fp = fopen(argv[1], "r");
    if (!fp)
    {
        perror("fopen");
        exit(EXIT_FAILURE);
    }

    // Run through
    char  *line = NULL;
    size_t len;
    int    one = 0;
    int    two = 0;
    while (getline(&line, &len, fp) != -1)
    {
        // Part 1
        int *row = tokenize(line, &len);
        if (safe(row, len))
            ++one;

        // Part 2
        for (size_t i = 0; i < len; ++i)
        {
            int *newrow = pop(row, len, i);
            if (safe(newrow, len - 1))
                ++two;
            free(newrow);
        }
        free(row);
    }

    printf("one: %d\n", one);
    printf("two: %d\n", two);

    fclose(fp);
    if (line)
        free(line);
    exit(EXIT_SUCCESS);
}

File: Python/2024/day02.py
"""
--- Day 2: Red-Nosed Reports ---
"""

from aocd import get_data

data = get_data(day=2, year=2024).splitlines()


def safe(i):
    """Checks whether row is safe."""
    sf = True
    flag = True
    inc = True
    for a, b in zip(i, i[1:]):
        if flag:
            inc = b - a > 0
        if (inc and a > b) or (not inc and a < b):
            sf = False
            break
        if abs(b - a) < 1 or abs(b - a) > 3:
            sf = False
            break
        flag = False
    return sf


ONE = 0
for ln in data:
    r = [int(x) for x in ln.split(" ")]
    if safe(r):
        ONE += 1
print("ONE:", ONE)

TWO = 0
for ln in data:
    r = [int(x) for x in ln.split(" ")]
    for idx in range(len(r)):
        nr = r[:idx] + r[idx + 1 :]
        if safe(nr):
            TWO += 1
            break
print("TWO:", TWO)
